File: kanaria_jvm/src/test/kotlin/com/kanaria/ExampleSentenceTest.kt
package com.kanaria

import com.kanaria.utils.AsciiUtils
import com.kanaria.utils.KanaUtils
import com.kanaria.utils.WidthUtils
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import java.nio.file.Files
import java.nio.file.Paths

class ExampleSentenceTest {
    @Test
    fun exampleSentence1() {
        val hankaku = "ﾁﾀﾀﾌﾟ ﾄﾃﾄﾃFoooo!!!11!"
        val zenkaku = "チタタプ　トテトテＦｏｏｏｏ！！！１１！"

        KanariaLoader.load()
        Assert.assertEquals(hankaku, UcsString.from(zenkaku).narrow().toString())
        Assert.assertEquals(zenkaku, UcsString.from(hankaku).wide().toString())
    }

    @Test
    fun exampleSentence2() {
        val hankaku = "吾輩ﾊ😺猫ﾃﾞｱﾙ😺"
        val zenkaku = "吾輩ハ😺猫デアル😺"

        KanariaLoader.load()
        Assert.assertEquals(hankaku, UcsString.from(zenkaku).narrow().toString())
        Assert.assertEquals(zenkaku, UcsString.from(hankaku).wide().toString())
    }

    @Test
    fun exampleSentence3() {
        val hankaku = "ﾌｼﾞｻﾝｺﾎﾟｫ"
        val zenkaku = "フジサンコポォ"

        KanariaLoader.load()
        Assert.assertEquals(hankaku, UcsString.from(zenkaku).narrow().toString())
        Assert.assertEquals(zenkaku, UcsString.from(hankaku).wide().toString())
    }

    @Test
    fun exampleSentence4() {
        val katakana = "チタタプ　トテトテ"
        val hiragana = "ちたたぷ　とてとて"

        KanariaLoader.load()
        Assert.assertEquals(katakana, UcsString.from(hiragana).katakana().toString())
        Assert.assertEquals(hiragana, UcsString.from(katakana).hiragana().toString())
    }

    @Test
    fun exampleSentence5() {
        val katakana = "吾輩ハ😺猫デアル😺"
        val hiragana = "吾輩は😺猫である😺"

        KanariaLoader.load()
        Assert.assertEquals(katakana, UcsString.from(hiragana).katakana().toString())
        Assert.assertEquals(hiragana, UcsString.from(katakana).hiragana().toString())
    }

    @Test
    fun exampleSentence6() {
        KanariaLoader.load()

        Assert.assertTrue(AsciiUtils.isAscii('a'))
        Assert.assertTrue(AsciiUtils.isWideAscii('ｗ'))
        Assert.assertTrue(AsciiUtils.isNarrowAscii('n'))
        Assert.assertFalse(AsciiUtils.isAscii('あ'))
        Assert.assertFalse(AsciiUtils.isWideAscii('漢'))
        Assert.assertFalse(AsciiUtils.isNarrowAscii('ｱ'))

        Assert.assertTrue(AsciiUtils.isLowerCase('a'))
        Assert.assertTrue(AsciiUtils.isWideLowerCase('ｗ'))
        Assert.assertTrue(AsciiUtils.isNarrowLowerCase('n'))
        Assert.assertFalse(AsciiUtils.isLowerCase('A'))
        Assert.assertFalse(AsciiUtils.isWideLowerCase('n'))
        Assert.assertFalse(AsciiUtils.isNarrowLowerCase('ｗ'))

        Assert.assertTrue(AsciiUtils.isUpperCase('A'))
        Assert.assertTrue(AsciiUtils.isWideUpperCase('Ｗ'))
        Assert.assertTrue(AsciiUtils.isNarrowUpperCase('N'))
        Assert.assertFalse(AsciiUtils.isUpperCase('a'))
        Assert.assertFalse(AsciiUtils.isWideUpperCase('ｗ'))
        Assert.assertFalse(AsciiUtils.isNarrowUpperCase('n'))

        Assert.assertTrue(AsciiUtils.isAsciiSymbol('*'))
        Assert.assertTrue(AsciiUtils.isWideAsciiSymbol('＠'))
        Assert.assertTrue(AsciiUtils.isNarrowAsciiSymbol('@'))
        Assert.assertFalse(AsciiUtils.isAsciiSymbol('、'))
        Assert.assertFalse(AsciiUtils.isWideAsciiSymbol('@'))
        Assert.assertFalse(AsciiUtils.isNarrowAsciiSymbol('＠'))

        Assert.assertTrue(AsciiUtils.isNumber('0'))
        Assert.assertTrue(AsciiUtils.isWideNumber('０'))
        Assert.assertTrue(AsciiUtils.isNarrowNumber('1'))
        Assert.assertFalse(AsciiUtils.isNumber('a'))
        Assert.assertFalse(AsciiUtils.isWideNumber('1'))
        Assert.assertFalse(AsciiUtils.isNarrowNumber('０'))

        Assert.assertEquals('a', AsciiUtils.convertToLowerCase('A'))
        Assert.assertEquals('A', AsciiUtils.convertToUpperCase('a'))
        Assert.assertEquals('あ', AsciiUtils.convertToLowerCase('あ'))
        Assert.assertEquals('0', AsciiUtils.convertToUpperCase('0'))
    }

    @Test
    fun exampleSentence7() {
        KanariaLoader.load()

        Assert.assertTrue(KanaUtils.isHiragana('あ'))
        Assert.assertFalse(KanaUtils.isHiragana('ア'))

        Assert.assertTrue(KanaUtils.isKatakana('ア'))
        Assert.assertTrue(KanaUtils.isWideKatakana('カ'))
        Assert.assertTrue(KanaUtils.isNarrowKatakana('ｻ'))
        Assert.assertFalse(KanaUtils.isKatakana('あ'))
        Assert.assertFalse(KanaUtils.isWideKatakana('ｶ'))
        Assert.assertFalse(KanaUtils.isNarrowKatakana('サ'))

        Assert.assertTrue(KanaUtils.isJisSymbol('ゟ'))
        Assert.assertTrue(KanaUtils.isWideJisSymbol('・'))
        Assert.assertTrue(KanaUtils.isNarrowJisSymbol('･'))
        Assert.assertFalse(KanaUtils.isJisSymbol('＠'))
        Assert.assertFalse(KanaUtils.isWideJisSymbol('･'))
        Assert.assertFalse(KanaUtils.isNarrowJisSymbol('・'))

        Assert.assertTrue(KanaUtils.isCanConvertHiragana('ワ'))
        Assert.assertFalse(KanaUtils.isCanConvertHiragana('ヷ'))

        Assert.assertEquals('あ', KanaUtils.convertToHiragana('ア'))
        Assert.assertEquals('ア', KanaUtils.convertToKatakana('あ'))
        Assert.assertEquals('ｱ', KanaUtils.convertToHiragana('ｱ'))
        Assert.assertEquals('漢', KanaUtils.convertToKatakana('漢'))
    }

    @Test
    fun exampleSentence8() {
        KanariaLoader.load()

        var wideResult: WidthUtils.ConvertToWideResult? = null

        wideResult = WidthUtils.convertToWide('ｱ', 0.toChar())
        Assert.assertEquals('ア', wideResult.result)
        Assert.assertFalse(wideResult.isPadding)

        wideResult = WidthUtils.convertToWide('ｶ', 'ﾞ')
        Assert.assertEquals('ガ', wideResult.result)
        Assert.assertTrue(wideResult.isPadding)

        wideResult = WidthUtils.convertToWide('ｶ', 'ﾟ')
        Assert.assertEquals('カ', wideResult.result)
        Assert.assertFalse(wideResult.isPadding)

        wideResult = WidthUtils.convertToWide('あ', 0.toChar())
        Assert.assertEquals('あ', wideResult.result)
        Assert.assertFalse(wideResult.isPadding)

        var narrowResult: WidthUtils.ConvertToNarrowResult? = null

        narrowResult = WidthUtils.convertToNarrow('ア')
        Assert.assertEquals('ｱ', narrowResult.first)
        Assert.assertNull(narrowResult.second)

        narrowResult = WidthUtils.convertToNarrow('ガ')
        Assert.assertEquals('ｶ', narrowResult.first)
        Assert.assertEquals('ﾞ', narrowResult.second)

        narrowResult = WidthUtils.convertToNarrow('パ')
        Assert.assertEquals('ﾊ', narrowResult.first)
        Assert.assertEquals('ﾟ', narrowResult.second)

        narrowResult = WidthUtils.convertToNarrow('あ')
        Assert.assertEquals('あ', narrowResult.first)
        Assert.assertNull(narrowResult.second)

        narrowResult = WidthUtils.convertToNarrow('ガ')
        Assert.assertEquals('ｶ', narrowResult.first)
        Assert.assertNotEquals(0.toChar(), narrowResult.second)
        Assert.assertNotEquals('ﾟ', narrowResult.second)
    }
}

File: kanaria_dotnet/KanariaTest/src/ExampleSentenceTest.cs
using Kanaria;
using Kanaria.Utils;
using NUnit.Framework;

namespace KanariaTest
{
    public class ExampleSentenceTest
    {
        [Test]
        public void ExampleSentence1()
        {
            var hankaku = "ﾁﾀﾀﾌﾟ ﾄﾃﾄﾃFoooo!!!11!";
            var zenkaku = "チタタプ　トテトテＦｏｏｏｏ！！！１１！";
            Assert.AreEqual(hankaku, UcsString.From(zenkaku).Narrow().ToString());
            Assert.AreEqual(zenkaku, UcsString.From(hankaku).Wide().ToString());
        }

        [Test]
        public void ExampleSentence2()
        {
            var hankaku = "吾輩ﾊ😺猫ﾃﾞｱﾙ😺";
            var zenkaku = "吾輩ハ😺猫デアル😺";
            Assert.AreEqual(hankaku, UcsString.From(zenkaku).Narrow().ToString());
            Assert.AreEqual(zenkaku, UcsString.From(hankaku).Wide().ToString());
        }

        [Test]
        public void ExampleSentence3()
        {
            var hankaku = "ﾌｼﾞｻﾝｺﾎﾟｫ";
            var zenkaku = "フジサンコポォ";
            Assert.AreEqual(hankaku, UcsString.From(zenkaku).Narrow().ToString());
            Assert.AreEqual(zenkaku, UcsString.From(hankaku).Wide().ToString());
        }


        [Test]
        public void ExampleSentence4()
        {
            var katakana = "チタタプ　トテトテ";
            var hiragana = "ちたたぷ　とてとて";
            Assert.AreEqual(katakana, UcsString.From(hiragana).Katakana().ToString());
            Assert.AreEqual(hiragana, UcsString.From(katakana).Hiragana().ToString());
        }

        [Test]
        public void ExampleSentence5()
        {
            var katakana = "吾輩ハ😺猫デアル😺";
            var hiragana = "吾輩は😺猫である😺";
            Assert.AreEqual(katakana, UcsString.From(hiragana).Katakana().ToString());
            Assert.AreEqual(hiragana, UcsString.From(katakana).Hiragana().ToString());
        }

        [Test]
        public void ExampleSentence6()
        {
            Assert.True(AsciiUtils.IsAscii('a'));
            Assert.True(AsciiUtils.IsWideAscii('ｗ'));
            Assert.True(AsciiUtils.IsNarrowAscii('n'));
            Assert.False(AsciiUtils.IsAscii('あ'));
            Assert.False(AsciiUtils.IsWideAscii('漢'));
            Assert.False(AsciiUtils.IsNarrowAscii('ｱ'));

            Assert.True(AsciiUtils.IsLowerCase('a'));
            Assert.True(AsciiUtils.IsWideLowerCase('ｗ'));
            Assert.True(AsciiUtils.IsNarrowLowerCase('n'));
            Assert.False(AsciiUtils.IsLowerCase('A'));
            Assert.False(AsciiUtils.IsWideLowerCase('n'));
            Assert.False(AsciiUtils.IsNarrowLowerCase('ｗ'));

            Assert.True(AsciiUtils.IsUpperCase('A'));
            Assert.True(AsciiUtils.IsWideUpperCase('Ｗ'));
            Assert.True(AsciiUtils.IsNarrowUpperCase('N'));
            Assert.False(AsciiUtils.IsUpperCase('a'));
            Assert.False(AsciiUtils.IsWideUpperCase('ｗ'));
            Assert.False(AsciiUtils.IsNarrowUpperCase('n'));

            Assert.True(AsciiUtils.IsAsciiSymbol('*'));
            Assert.True(AsciiUtils.IsWideAsciiSymbol('＠'));
            Assert.True(AsciiUtils.IsNarrowAsciiSymbol('@'));
            Assert.False(AsciiUtils.IsAsciiSymbol('、'));
            Assert.False(AsciiUtils.IsWideAsciiSymbol('@'));
            Assert.False(AsciiUtils.IsNarrowAsciiSymbol('＠'));

            Assert.True(AsciiUtils.IsNumber('0'));
            Assert.True(AsciiUtils.IsWideNumber('０'));
            Assert.True(AsciiUtils.IsNarrowNumber('1'));
            Assert.False(AsciiUtils.IsNumber('a'));
            Assert.False(AsciiUtils.IsWideNumber('1'));
            Assert.False(AsciiUtils.IsNarrowNumber('０'));

            Assert.AreEqual('a', AsciiUtils.ConvertToLowerCase('A'));
            Assert.AreEqual('A', AsciiUtils.ConvertToUpperCase('a'));
            Assert.AreEqual('あ', AsciiUtils.ConvertToLowerCase('あ'));
            Assert.AreEqual('0', AsciiUtils.ConvertToUpperCase('0'));
        }

        [Test]
        public void ExampleSentence7()
        {
            Assert.True(KanaUtils.IsHiragana('あ'));
            Assert.False(KanaUtils.IsHiragana('ア'));

            Assert.True(KanaUtils.IsKatakana('ア'));
            Assert.True(KanaUtils.IsWideKatakana('カ'));
            Assert.True(KanaUtils.IsNarrowKatakana('ｻ'));
            Assert.False(KanaUtils.IsKatakana('あ'));
            Assert.False(KanaUtils.IsWideKatakana('ｶ'));
            Assert.False(KanaUtils.IsNarrowKatakana('サ'));

            Assert.True(KanaUtils.IsJisSymbol('ゟ'));
            Assert.True(KanaUtils.IsWideJisSymbol('・'));
            Assert.True(KanaUtils.IsNarrowJisSymbol('･'));
            Assert.False(KanaUtils.IsJisSymbol('＠'));
            Assert.False(KanaUtils.IsWideJisSymbol('･'));
            Assert.False(KanaUtils.IsNarrowJisSymbol('・'));

            Assert.True(KanaUtils.IsCanShiftToHiragana('ワ'));
            Assert.False(KanaUtils.IsCanShiftToHiragana('ヷ'));

            Assert.AreEqual('あ', KanaUtils.ConvertToHiragana('ア'));
            Assert.AreEqual('ア', KanaUtils.ConvertToKatakana('あ'));
            Assert.AreEqual('ｱ', KanaUtils.ConvertToHiragana('ｱ'));
            Assert.AreEqual('漢', KanaUtils.ConvertToKatakana('漢'));
        }

        [Test]
        public void ExampleSentence8()
        {
            var isPad = false;
            
            Assert.AreEqual('ア', WidthUtils.ConvertToWide('ｱ', (char) 0, out isPad));
            Assert.False(isPad);
            
            Assert.AreEqual('ガ', WidthUtils.ConvertToWide('ｶ', 'ﾞ', out isPad));
            Assert.True(isPad);
            
            Assert.AreEqual('カ', WidthUtils.ConvertToWide('ｶ', 'ﾟ', out isPad));
            Assert.False(isPad);
            
            Assert.AreEqual('あ', WidthUtils.ConvertToWide('あ', (char) 0, out isPad));
            Assert.False(isPad);

            var second = char.MaxValue;
            
            Assert.AreEqual('ｱ', WidthUtils.ConvertToNarrow('ア', out second));
            Assert.AreEqual((char) 0, second);
            
            Assert.AreEqual('ｶ', WidthUtils.ConvertToNarrow('ガ', out second));
            Assert.AreEqual('ﾞ', second);
            
            Assert.AreEqual('ﾊ', WidthUtils.ConvertToNarrow('パ', out second));
            Assert.AreEqual('ﾟ', second);
            
            Assert.AreEqual('あ', WidthUtils.ConvertToNarrow('あ', out second));
            Assert.AreEqual((char) 0, second);
            
            Assert.AreEqual('ｶ', WidthUtils.ConvertToNarrow('ガ', out second));
            Assert.AreNotEqual((char) 0, second);
            Assert.AreNotEqual('ﾟ', second);
        }
        
        [Test]
            public void ExampleSentence9() {
            var source = "吾輩は😺猫である😺";
            var expect = "吾輩ﾊ😺猫ﾃﾞｱﾙ😺";
            var expect2 = "吾輩ハ😺猫デアル😺";

            Assert.AreEqual(expect, UcsString.From(source).Katakana().Narrow(ConvertTarget.All).ToString());
            Assert.AreEqual(expect2, UcsString.From(source).Katakana().Narrow(ConvertTarget.Number | ConvertTarget.Symbol | ConvertTarget.Alphabet).ToString());
        }
    }
}